### PR TITLE
Refactorized permissions api to use promises as the standard w3c spec

### LINKED ISSUE #120: Return a Promise in Push.Permission.request instead of using callbacks
Maybe we can return a promise here in order to have an implementation more similar to the native solution instead of using callbacks.

https://github.com/Nickersoft/push.js/blob/c990321f90e1a205cafadc03809018cba7692c36/src/classes/Permission.js#L44

I can make a PR if you agree with this but it will need some triage to see which version should adpot this kind of improvements.

### PR DESCRIPTION
**Pull request by basiclines**
Fixes #120 
I've refactorized the code related with `Push.Permission.request()` in order to use Promises instead of callbacks. All related tests have been fixed in order to support this change.

This currently breaks the old behavior in Push.js. Should we keep the callbacks support?

### CODE REVIEW

**Comment by Nickersoft** [comment]
Yeah, I'd say so. Don't want to break anyone's current implementation. I'd say the safest thing to do is depreciate the current method signature ([see here](http://usejsdoc.org/tags-deprecated.html)) just to let people know they should change their implementation. However, you can't overload methods in JavaScript so I'd look an implementation [such as this one](https://www.bennadel.com/blog/2008-overloading-javascript-functions-using-a-sub-function-approach.htm) to support both signatures.

**Comment by basiclines** [comment]
I've decided to create 2 new methods (`_requestWithCallback` and `_requestAsPromise`) using de facto standard of `_` which implies those are internal methods for the library. In this way, we get a cleaner code than creating 2 inner functions inside the main `request` method.

Maybe i should maintain tests for both implementations, but i'm not sure about it, as the callback implementation is deprecated and there shouldn't be any future changes affecting it.

File: src/classes/Permission.js [hunk 1]
```diff
@@ -45,6 +56,41 @@ export default class Permission {
...
+
+    return new Promise((resolvePromise, rejectPromise) => {
+
+      var resolver = result => (isGranted(result)) ? resolvePromise() : rejectPromise();
+
+      if (hasPermissions) {
+       resolver(existing)
+      }
+      else if (isModernAPI) {
+        this._win.Notification.requestPermission().then(result => { resolver(result) }).catch(rejectPromise)
```
Comment (n3il) [inline src/classes/Permission.js]: Wouldn't this run into the same issue as described in https://github.com/Nickersoft/push.js/issues/117?

The `.then()` implementation is the issue with Safari.

Comment (basiclines) [reply src/classes/Permission.js]: I would take a look then and try to fix it. Anyway i should create a separated PR to resolve this specific issue.

Comment (n3il) [reply src/classes/Permission.js]: I'm looking into the fix as well. It was actually me who mistakenly linked the `.then` ticket to this PR. A separate PR does make sense. What you have looks good though

**Comment by basiclines** [comment]
I've added the old test back while keeping the new ones for the promise based API, mostly to keep codeclimate coverage all up.

**Comment by Nickersoft** [comment]
@basiclines Could you see if you could get the coverage back up? It's okay if there are some things that are hard to test, but 16.9% seems to be a hard hit to the code coverage.

**Comment by basiclines** [comment]
I have to rebase the branch first. About the code coverage, i don't really know why has dropped so much, when actually i've committed new tests to the branch.

**Comment by basiclines** [comment]
All fixed and running clean @Nickersoft 😄

**Comment by Nickersoft** [comment]
I got you :) Thanks for the contribution!

**Pull request merged**